java/adapter/jdbc/src/main/java/org/apache/arrow/adapter/jdbc/JdbcToArrowUtils.java

```diff
@@ -93,119 +98,229 @@
+   *   <li>DATE --> ArrowType.Date(DateUnit.MILLISECOND)</li>
+   *   <li>TIME --> ArrowType.Time(TimeUnit.MILLISECOND, 32)</li>
+   *   <li>TIMESTAMP --> ArrowType.Timestamp(TimeUnit.MILLISECOND, calendar timezone)</li>
+   *   <li>CLOB --> ArrowType.Utf8</li>
+   *   <li>BLOB --> ArrowType.Binary</li>
+   * </ul>
+   *
+   * @param fieldInfo The field information to construct the <code>ArrowType</code> from.
+   * @param calendar The calendar to use when constructing the <code>ArrowType.Timestamp</code>
+   *                 for {@link java.sql.Types#TIMESTAMP} types.
+   * @return The corresponding <code>ArrowType</code>.
+   * @throws NullPointerException if either <code>fieldInfo</code> or <code>calendar</code> are <code>null</code>.
+   */
+  public static ArrowType getArrowTypeForJdbcField(JdbcFieldInfo fieldInfo, Calendar calendar) {
+    Preconditions.checkNotNull(fieldInfo, "JdbcFieldInfo object cannot be null");
+    Preconditions.checkNotNull(calendar, "Calendar object cannot be null");
+
+    final ArrowType arrowType;
+
+    switch (fieldInfo.getJdbcType()) {
+      case Types.BOOLEAN:
+      case Types.BIT:
+        arrowType = new ArrowType.Bool();
+        break;
+      case Types.TINYINT:
+        arrowType = new ArrowType.Int(8, true);
+        break;
+      case Types.SMALLINT:
+        arrowType = new ArrowType.Int(16, true);
+        break;
+      case Types.INTEGER:
+        arrowType = new ArrowType.Int(32, true);
+        break;
+      case Types.BIGINT:
+        arrowType = new ArrowType.Int(64, true);
+        break;
+      case Types.NUMERIC:
+      case Types.DECIMAL:
+        int precision = fieldInfo.getPrecision();
+        int scale = fieldInfo.getScale();
+        arrowType = new ArrowType.Decimal(precision, scale);
+        break;
+      case Types.REAL:
+      case Types.FLOAT:
+        arrowType = new ArrowType.FloatingPoint(SINGLE);
+        break;
+      case Types.DOUBLE:
+        arrowType = new ArrowType.FloatingPoint(DOUBLE);
+        break;
+      case Types.CHAR:
+      case Types.NCHAR:
+      case Types.VARCHAR:
+      case Types.NVARCHAR:
+      case Types.LONGVARCHAR:
+      case Types.LONGNVARCHAR:
+      case Types.CLOB:
+        arrowType = new ArrowType.Utf8();
+        break;
+      case Types.DATE:
+        arrowType = new ArrowType.Date(DateUnit.MILLISECOND);
+        break;
+      case Types.TIME:
+        arrowType = new ArrowType.Time(TimeUnit.MILLISECOND, 32);
+        break;
+      case Types.TIMESTAMP:
+        arrowType = new ArrowType.Timestamp(TimeUnit.MILLISECOND, calendar.getTimeZone().getID());
+        break;
+      case Types.BINARY:
+      case Types.VARBINARY:
+      case Types.LONGVARBINARY:
+      case Types.BLOB:
+        arrowType = new ArrowType.Binary();
+        break;
+      case Types.ARRAY:
+        arrowType = new ArrowType.List();
+        break;
+      default:
+        // no-op, shouldn't get here
+        arrowType = null;
```

Comment:
would it make sense to throw an exception instead?

Reply:
or at least log?

Reply:
I wasn't trying to stray too far from what the original author was doing, since his work did pass code review ...

The diff is a little confusing, but the original logic is on the old logic is on the old (red) line 249.